provider/openstack/networking.go

```diff
@@ -487,25 +490,37 @@ func (n *NeutronNetworking) NetworkInterfaces(instanceIDs []instance.Id) ([]core
 		subnetIDToCIDR[sub.ProviderId.String()] = sub.CIDR
 	}
 
-	// Unfortunately the client does not support filter by anything other
-	// than tags so we need to grab everything and apply the filter here.
 	neutronClient := n.neutron()
-	allInstPorts, err := neutronClient.ListPortsV2()
+	filter := projectIDFilter(n.client().TenantId())
+
+	fips, err := neutronClient.ListFloatingIPsV2(filter)
+	if err != nil {
+		return nil, errors.Annotate(err, "listing floating IPs")
+	}
+
+	// Map private IP to public IP for every assigned FIP.
+	fixedToFIP := make(map[string]string)
+	for _, fip := range fips {
+		if fip.FixedIP == "" {
+			continue
+		}
+		fixedToFIP[fip.FixedIP] = fip.IP
+	}
+
+	allInstPorts, err := neutronClient.ListPortsV2(filter)
```

Comment:
Adding the filter here is a really good optimization, as we don't then need to filter in juju directly.